proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,634 @@
+this provides an accurate representation of the current state of the
+Haskell module system and we describe our changes with respect to it. [#]_
+
+.. [#] I find the relational setting very amenable for this work, since it makes it
+       easy to describe the rules while gracefully handling un- or multiply-defined
+       bindings.
+
+The text here is mostly expository. There is a `Datalog implementation`_ of some
+of the rules.
+
+The proposed changes are:
+
+1. `Modules as entities`_
+2. `Module export specifiers`_
+3. `Local modules`_
+4. `Module aliases`_
+
+Terminology
+~~~~~~~~~~~
+
+-  A *compilation unit* is the unit of Haskell compilation, currently a
+   single ``.hs`` file.
+-  A *file module* is a module which is defined at the top level of a
+   ``.hs`` file. Currently this corresponds to a compilation unit.
+-  A *module* is a named scope, of which file modules are currently the
+   only examples.
+-  A *simple name* is a name that is given to something when it is
+   defined.
+-  A *qualified name* is a more complex form of name which can *refer*
+   to an entity. We will say more about qualified names later.
+-  A *dotted name* is a (simple) module name that contains ``.``\ s, e.g
+   ``Data.Set``. We will want to distinguish these from qualified names.
+-  An *entity* is something that can be referred to with a name.
+   Currently this means types, values, field accessors, etc.
+-  A *binding* is a mapping from a name to an entity.
+
+Modules as entities
+~~~~~~~~~~~~~~~~~~~
+
+This change makes modules into entities, and allows them to be referred
+to. However, the only place that they can be referred to is in the
+qualifiers of qualified names, so this changes little.
+
+I believe this would be entirely internal, and would have no effect on
+what names resolve to in existing Haskell programs. As such, it could be
+implemented without a language extension flag, and is completely
+compatible with existing code.
+
+Modules become entities
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Modules become entities. All scopes now have a *module namespace*, which
+contains the name bindings for modules. [#]_
+
+.. [#] I will ignore namespaces for the rest of this proposal. They don't
+       interact with this proposal much, since it is easy to distinguish which
+       bindings belong to which namespaces by classifying the *entities* according
+       to which namespace they belong to. Then we can have a "unified" namespace and
+       simply filter out bindings that refer to the wrong kind of entities. 
+
+Qualified names become more structured
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Currently, qualified names consist of a simple name, possibly qualified
+by a module name. That is, they could be represented by a type like:
+
+::
+
+   data QualName = Qualified ModuleName Name | Simple Name
+
+We propose to replace this with a type like:
+
+::
+
+   data QualName = Qualified QualName Name | Simple Name
```

Comment:
Did you consider the "right-recursive" rather than "left-recursive" 
```haskell
data QualName = Qualified Name QualName | Simple Name
```

I think this makes `refersTo` a bit clearer, as it's now tail recursive, and frankly, all things equal, left-to-right is easier given that's the way Haskell text flows.

Reply:
Interesting, I hadn't thought of that. The current way makes more sense to me as an evaluator (evaluate the qualifier and then access it), but in the end it's an implementation detail.

However, there is one quirk. I realise that subconsciously I was preparing myself to add a `DottedName ....` alternative. In which case we can no longer always assume that the first segment is going to be an independent `Name`, whereas we can *always* assume that the last segment is.

Reply:
Oh actually I completely take back my suggestion. `.` would be right-associative if we had other expression forms we had to disambiguate with, so your way is right.